Document how to get around storage-limitations that some installations may encounter (#60)
plus3it/nessus-agent-formula: docs/StorageUsage-Linux.md

# Background

Some recent (started receiving reports in late 2025) users of this formula have reported that, after this formula has installed and configured their Nessus Agents on hosts, they get scan-failures on their Linux hosts. When they check their Nessus servers' logs, they find error-log entries indicating that the failures are due to storage-space exhaustion.

Initial investigation of this issue found that what was happening was that the storage-space problem was a result of the scan process installing plugins. The initial issue was exhaustion caused by the scanner staging and then attempting to unarchive to-be-installed plugins in the `/opt/nessus_agent/var/tmp` directory. Attempts to forestall this problem via updates to this project's automation solved _that_ problem. However, it didn't solve the overarching space-exhaustion issues.

> Further testing showed that fixing the inital probelem caused by the `/opt/nessus_agent/var/tmp` directory filling up then resulted in the `/opt/nessus_agent/var/nessus/` directory filling up. Initially, it was assumed that this was due to the _plugins_ successful installation. However, these plugins only accounted for a further 10MiB of space-consumption (in the `/opt/nessus_agent/var/nessus//plugins` subdirectory). Once a scan started to run, it begins to generate a collection of `*.db` and `*.db-*` files (directly) within the `/opt/nessus_agent/var/nessus/` directory. On a testing-system with the `/` filesystem over-provisioned (in testing, grew the `/` filesystem by 10GiB), it was found that the _first_ successful scan of a system resulted in just shy of 1.8GiB of persistent `*.db` and `*.db-*` files being created.

Reviewer: ```suggestion
Further testing showed that fixing the initial problem caused by the `/opt/nessus_agent/var/tmp` directory filling up then resulted in the `/opt/nessus_agent/var/nessus/` directory filling up. Initially, it was assumed that this was due to the _plugins_ successful installation. However, these plugins only accounted for a further 10MiB of space-consumption (in the `/opt/nessus_agent/var/nessus//plugins` subdirectory). Once a scan started to run, it begins to generate a collection of `*.db` and `*.db-*` files (directly) within the `/opt/nessus_agent/var/nessus/` directory. On a testing-system with the `/` filesystem over-provisioned (in testing, grew the `/` filesystem by 10GiB), it was found that the _first_ successful scan of a system resulted in just shy of 1.8GiB of persistent `*.db` and `*.db-*` files being created.
```

Fixing this configuration requires knowing:

* The device-path of the block-device hosting the LVM2 volume-group
* The name of the LVM2 volume-group hosting the root volumes and filesystems
* The name of logical volume hosting the `filesystem`.

While it's not _strictly_ necessary to know these values, _not_ knowing them considerably increases the complexity of the necessary userData payload[^1].

With the `cloud-init` utility present in the host operating system, the problem can be prevented with a mixed-MIME userData payload. Adjust the following userData payload to reflect your host's _actual_ LVM2 volume-group name, name of the volume hosting the `/` parition and the device-path of the block-device hosting the root filesystem(s):

```
Content-Type: multipart/mixed; boundary="===============BOUNDARY=="
MIME-Version: 1.0
Number-Attachments: 2

--===============BOUNDARY==
Content-Disposition: attachment; filename="cloud.cfg"
Content-Transfer-Encoding: 7bit
Content-Type: text/cloud-config
Mime-Version: 1.0

#cloud-config

# Ensure that we use any available extra storage
growpart:
  mode: auto

>   devices: [
>       '/dev/nvme0n1p4'
>     ]

Reviewer: It seems quite strange to mix yaml syntax with json syntax within the same file... works, but more confusing...

```suggestion
  devices:
    - /dev/nvme0n1p4
```

Author: I want to leave this as specified. Normally, when launching systems, I have it as:
```
growpart:
  mode: auto
  devices: [
      '/dev/xvda',
      '/dev/xvda1',
      '/dev/xvda2',
      '/dev/xvda3',
      '/dev/xvda4',
      '/dev/nvme0n1',
      '/dev/nvme0n1p1',
      '/dev/nvme0n1p2',
      '/dev/nvme0n1p3',
      '/dev/nvme0n1p4'
    ]
```
And, while I understand that 
```yaml
growpart:
  mode: auto
  devices:
    - /dev/path1
    - /dev/path2
    - …
    - /dev/pathN
```
_Should_ also work, the `[ ]` specification-method is what came from the [cloud-init documentation](https://docs.cloud-init.io/en/latest/reference/yaml_examples/disk_setup.html#grow-partitions).

--===============BOUNDARY==
MIME-Version: 1.0
Content-Type: text/x-shellscript
Content-Disposition: attachment; filename="00_bootdev_setup.sh"
#!/bin/bash
set -euo pipefail
#

> # UserData scriptlet to extent rootVol LVM2 volume

Reviewer: ```suggestion
# UserData scriptlet to extend rootVol LVM2 volume
```

#
#################################################################

lvextend -rl +100%FREE RootVG/rootVol

--===============BOUNDARY==--
```

The above userData will

1. Attempt to grow the disk-partition known to the system as `/dev/nvme0n1p4`.

> 2. Attempt to grow the LVM2 volume-object (that lives on the `/dev/nvme0n1p4` block-device node)

Reviewer: for consistency with prior bullet...

```suggestion
2. Attempt to grow the LVM2 volume-object (that lives on the `/dev/nvme0n1p4` block-device node).
```

```
[...ELIDED...]

--===============BOUNDARY==
MIME-Version: 1.0
Content-Type: text/x-shellscript
Content-Disposition: attachment; filename="01_nessusdev_setup.sh"
#!/bin/bash
set -euo pipefail
#
# UserData scriptlet create a dedicated LVM volume mount for the
# Nessus agent
#
#################################################################

lvcreate -L <SIZE_OF_VOLUME> -n nessusVol RootVG
mkfs -t xfs -L NessusAgent /dev/RootVG/nessusVol
install -dDm 0700 /opt/nessus_agent
print -f '/dev/RootVG/nessusVol\t/opt/nessus_agent\txfs\tdefaults\t1 1\n' >> /etc/fstab
systemctl daemon-reload
mount /dev/RootVG/nessusVol

--===============BOUNDARY==--
```

Can be used. The above:

1. Creates an LVM2 volume named "`nessusVol`" within the "`RootVG`" LVM2 volume-group
1. Creates an XFS filesystem, with the "`NessusAgent`" label, on top of the "`nessusVol`" LVM2 volume
1. Creates the mountpoint, "`/opt/nessus_agent`"
1. Adds a suitable entry to the system's "`/etc/fstab`" file
1. Makes `systemd` reread the "`/etc/fstab`" file to update its mount-unit definitions
1. Mounts the XFS-formatted filesystem hosted on "`/dev/RootVG/nessusVol`" to the "`/opt/nessus_agent`" mountpoint

> [^1]: The userData-payload shown is _very_ simplified and not very flexible as a result. Flexibility can be achieved by using "discovery" logic to determing things like the name of LVM2 volume-group and volume-name associated with the `/` filesystem.

Reviewer: ```suggestion
[^1]: The userData-payload shown is _very_ simplified and not very flexible as a result. Flexibility can be achieved by using "discovery" logic to determine things like the name of LVM2 volume-group and volume-name associated with the `/` filesystem.
```
